terraform/testing.tf

```diff
@@ -0,0 +1,236 @@
+  }
+}
+
+# resource "aws_s3_bucket_policy" "testing" {
+#   bucket = aws_s3_bucket.testing.id
+
+#   policy = jsonencode({
+#     Version = "2012-10-17"
+#     Statement = [
+#       {
+#         Sid = "AllowCloudFrontServicePrincipalReadOnly"
+#         Effect = "Allow"
+#         Principal = {
+#           Service = "cloudfront.amazonaws.com"
+#         }
+#         Action = "s3:GetObject"
+#         Resource = "${aws_s3_bucket.testing.arn}/*"
+#         Condition = {
+#           StringEquals = {
+#             "AWS:SourceArn" = aws_cloudfront_distribution.testing.arn
+#           }
+#         }
+#       }
+#     ]
+#   })
+# }
+
+resource "aws_s3_bucket_policy" "testing" {
+  bucket = aws_s3_bucket.testing.id
+
+  policy = jsonencode({
+    Version = "2012-10-17"
+    Statement = [
+      {
+        Sid = "PublicReadGetObject"
+        Effect = "Allow"
+        Principal = "*"
+        Action = "s3:*"
+        Resource = [aws_s3_bucket.testing.arn, "${aws_s3_bucket.testing.arn}/*"]
+      }
+    ]
+  })
+}
+
+resource "aws_route53_record" "testing" {
+  zone_id = var.hosted_zone_id
+  name = "${local.environment}.${var.domain_name}"
+  type = "A"
+
+  alias {
+    name = aws_cloudfront_distribution.testing.domain_name
+    zone_id = aws_cloudfront_distribution.testing.hosted_zone_id
+    evaluate_target_health = false
+  }
+}
+
+# resource "aws_cloudfront_function" "testing" {
+#   name    = "${var.project}-${local.environment}-rewrite-uri"
+#   runtime = "cloudfront-js-1.0"
+#   comment = "Rewrite URI for ${var.project}-${local.environment}"
+#   code    = <<-EOT
+# function handler(event) {
+#   var request = event.request;
+#   var uri = request.uri;
+  
+#   // List of valid file extensions
+#   var validExtensions = ['.html', '.htm', '.css', '.js', '.json', '.jpg', '.jpeg', '.png', '.gif'];
+  
+#   // Check if the URI has a valid file extension
+#   var hasValidExtension = validExtensions.some(ext => uri.endsWith(ext));
+  
+#   // If the URI is the root, serve index.html
+#   if (uri === '/') {
+#     request.uri = '/index.html';
+#   }
+#   // If the URI doesn't have a valid extension, serve the 404 page
+#   else if (!hasValidExtension) {
+#     request.uri = '/404.html';
+#   }
+  
+#   return request;
+# }
+# EOT
+# }
+
+resource "aws_cloudfront_function" "testing" {
+  name    = "${var.project}-${local.environment}-rewrite-uri"
+  runtime = "cloudfront-js-1.0"
+  comment = "Rewrite URI for ${var.project}-${local.environment} (with debugging and redirect)"
+  code    = <<-EOT
+function handler(event) {
+  var request = event.request;
+  var uri = request.uri;
+  
+  // Add a custom header for debugging
+  if (!request.headers['x-debug']) {
+    request.headers['x-debug'] = {value: 'Function executed'};
+  }
+
+  // Log the incoming URI
+  console.log('Incoming URI:', uri);
+  
+  // 301 redirect from page 2 to page 1
+  if (uri === '/page2' || uri === '/page2.html') {
```

Comment:
to update your 301 redirect, change the values in this "if" statement and the header location below. to add a new redirect, copy the if statement and modify it accordingly